fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/transaction/TransactionImpl.java

```diff
@@ -0,0 +1,188 @@
+	private static final ArrayList<TransactionImpl> STACK = new ArrayList<>();
+	// The current transaction depth.
+	private static int currentDepth = -1;
+
+	private static void acquireLock() {
+		if (Thread.currentThread() != serverThread) outerLock.lock();
+		innerLock.lock();
+	}
+
+	private static void releaseLock() {
+		innerLock.unlock();
+		if (Thread.currentThread() != serverThread) outerLock.unlock();
+	}
+
+	public static Transaction openOuter() {
+		if (isOpen()) {
+			throw new IllegalStateException("An outer transaction is already active on this thread.");
+		}
+
+		acquireLock();
+
+		if (STACK.size() == 0) {
+			STACK.add(new TransactionImpl(0));
+		}
+
+		currentDepth = 0;
+		TransactionImpl current = STACK.get(currentDepth);
+		current.isOpen = true;
+		return current;
+	}
+
+	public static boolean isOpen() {
+		return innerLock.isHeldByCurrentThread();
+	}
+
+	public static void setServerThread(Thread serverThread) {
+		if (innerLock.isLocked()) {
+			throw new AssertionError("Trying to change server thread while a transaction is open.");
+		}
+
+		TransactionImpl.serverThread = serverThread;
+	}
+
+	TransactionImpl(int nestingDepth) {
+		this.nestingDepth = nestingDepth;
+	}
+
+	private final int nestingDepth;
+	private final List<CloseCallback> closeCallbacks = new ArrayList<>();
+	private boolean isOpen = false;
+
+	// Validate that this is the correct thread for transaction operations.
+	private static void validateCurrentThread() {
+		if (!innerLock.isHeldByCurrentThread()) {
+			throw new IllegalStateException(String.format(
+					"Transaction operations are already ongoing on another thread.\nCurrent thread: %s.\n",
+					Thread.currentThread().getName()));
+		}
+	}
+
+	// Validate that this transaction is the current one, and that this is called on the correct thread.
+	private void validateCurrentTransaction() {
+		validateCurrentThread();
+
+		if (STACK.get(currentDepth) != this) {
+			throw new IllegalStateException("Transaction operation can only be applied to the current transaction.");
+		}
+	}
+
+	// Validate that this transaction is open.
+	private void validateOpen() {
+		if (!isOpen) {
+			throw new IllegalStateException("Transaction operation cannot be applied to a closed transaction.");
+		}
+	}
+
+	@Override
+	public Transaction openNested() {
+		validateCurrentTransaction();
+		validateOpen();
+		currentDepth++;
+
+		if (currentDepth == STACK.size()) {
+			STACK.add(new TransactionImpl(currentDepth));
+		}
+
+		TransactionImpl current = STACK.get(currentDepth);
+		current.isOpen = true;
+		return current;
+	}
+
+	private void close(Result result) {
+		validateCurrentTransaction();
+		validateOpen();
+		// block transaction operations
+		isOpen = false;
+
+		// invoke callbacks in reverse order
+		for (int i = closeCallbacks.size()-1; i >= 0; i--) {
+			closeCallbacks.get(i).onClose(this, result);
```

Comment:
Totally a nitpick, but again I'd suggest using a for-each loop.
I.e. Change the ArrayList to a LinkedList and change addCloseCallback to use addFirst().
I think this would come mostly down to preference, so feel free to just ignore this.

Reply:
No, `LinkedList`s are much worse performance-wise and will need an allocation for every insertion.

Reply:
Yeah, I don't like LinkedLists either. But I'm also not a huge fan of indexing loops where they aren't necessary.
Another option would be using an ArrayDeque and the descendingIterator.
But again, this is absolutely fine as is and I wouldn't worry about it too much.

Reply:
That style of loop comes from a C idiom, it works well for counting down to 0 for unsigned data types. It's uncommon enough in the Java world to be confusing, and I personally hate the formatting of the arrow. Just stick to `i = size - 1; i >= 0; i--` as it's the clearest and most idiomatic in Java.

Reply:
Yup, it was just a little joke :)